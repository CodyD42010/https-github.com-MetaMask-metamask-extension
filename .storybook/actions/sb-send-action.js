export const updateSendStatus = (status) => ({
  type: 'send/updateSendStatus',
  payload: status,
});

export const updateAmountMode = (mode) => ({
  type: 'send/updateAmountMode',
  payload: mode,
});
