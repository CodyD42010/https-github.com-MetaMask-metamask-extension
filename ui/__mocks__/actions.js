const actions = require('../store/actions');

const ERC20_TOKEN_1_MOCK = '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599'; // WBTC
const ERC20_TOKEN_2_MOCK = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'; // USDC
const mockTokenDetails = {
  [ERC20_TOKEN_1_MOCK]: {
    address: ERC20_TOKEN_1_MOCK,
    standard: 'ERC20',
    decimals: 8,
    name: 'Wrapped Bitcoin',
  },
  [ERC20_TOKEN_2_MOCK]: {
    address: ERC20_TOKEN_2_MOCK,
    standard: 'ERC20',
    decimals: 6,
    name: 'USD Coin',
  },
};

module.exports = {
  ...actions,
  getTokenStandardAndDetails: (address) => {
    if (!mockTokenDetails[address]) {
      return Promise.reject(new Error('Token not found'));
    }
    return Promise.resolve(mockTokenDetails[address]);
  },
  // eslint-disable-next-line no-empty-function
  trackMetaMetricsEvent: () => {},
};
