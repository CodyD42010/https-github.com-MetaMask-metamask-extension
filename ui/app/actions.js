var actions = {
  GO_HOME: 'GO_HOME',
  goHome: goHome,
  // menu state
  getNetworkStatus: 'getNetworkStatus',

  // remote state
  UPDATE_METAMASK_STATE: 'UPDATE_METAMASK_STATE',
  updateMetamaskState: updateMetamaskState,
  // intialize screen
  AGREE_TO_DISCLAIMER: 'AGREE_TO_DISCLAIMER',
  agreeToDisclaimer: agreeToDisclaimer,
  CREATE_NEW_VAULT_IN_PROGRESS: 'CREATE_NEW_VAULT_IN_PROGRESS',
  SHOW_CREATE_VAULT: 'SHOW_CREATE_VAULT',
  SHOW_RESTORE_VAULT: 'SHOW_RESTORE_VAULT',
  SHOW_INIT_MENU: 'SHOW_INIT_MENU',
  SHOW_NEW_VAULT_SEED: 'SHOW_NEW_VAULT_SEED',
  SHOW_INFO_PAGE: 'SHOW_INFO_PAGE',
  RECOVER_FROM_SEED: 'RECOVER_FROM_SEED',
  CLEAR_SEED_WORD_CACHE: 'CLEAR_SEED_WORD_CACHE',
  clearSeedWordCache: clearSeedWordCache,
  recoverFromSeed: recoverFromSeed,
  unlockMetamask: unlockMetamask,
  unlockFailed: unlockFailed,
  showCreateVault: showCreateVault,
  showRestoreVault: showRestoreVault,
  showInitializeMenu: showInitializeMenu,
  createNewVault: createNewVault,
  createNewVaultInProgress: createNewVaultInProgress,
  showNewVaultSeed: showNewVaultSeed,
  showInfoPage: showInfoPage,
  // seed recovery actions
  REVEAL_SEED_CONFIRMATION: 'REVEAL_SEED_CONFIRMATION',
  revealSeedConfirmation: revealSeedConfirmation,
  requestRevealSeed: requestRevealSeed,
  // unlock screen
  UNLOCK_IN_PROGRESS: 'UNLOCK_IN_PROGRESS',
  UNLOCK_FAILED: 'UNLOCK_FAILED',
  UNLOCK_METAMASK: 'UNLOCK_METAMASK',
  LOCK_METAMASK: 'LOCK_METAMASK',
  tryUnlockMetamask: tryUnlockMetamask,
  lockMetamask: lockMetamask,
  unlockInProgress: unlockInProgress,
  // error handling
  displayWarning: displayWarning,
  showWarning: showWarning, // alias
  DISPLAY_WARNING: 'DISPLAY_WARNING',
  HIDE_WARNING: 'HIDE_WARNING',
  hideWarning: hideWarning,
  // accounts screen
  SET_SELECTED_ACCOUNT: 'SET_SELECTED_ACCOUNT',
  SHOW_ACCOUNT_DETAIL: 'SHOW_ACCOUNT_DETAIL',
  SHOW_ACCOUNTS_PAGE: 'SHOW_ACCOUNTS_PAGE',
  SHOW_CONF_TX_PAGE: 'SHOW_CONF_TX_PAGE',
  SHOW_CONF_MSG_PAGE: 'SHOW_CONF_MSG_PAGE',
  REVEAL_ACCOUNT: 'REVEAL_ACCOUNT',
  revealAccount: revealAccount,
  // account detail screen
  SHOW_SEND_PAGE: 'SHOW_SEND_PAGE',
  showSendPage: showSendPage,
  REQUEST_ACCOUNT_EXPORT: 'REQUEST_ACCOUNT_EXPORT',
  requestExportAccount: requestExportAccount,
  EXPORT_ACCOUNT: 'EXPORT_ACCOUNT',
  exportAccount: exportAccount,
  SHOW_PRIVATE_KEY: 'SHOW_PRIVATE_KEY',
  showPrivateKey: showPrivateKey,
  SAVE_ACCOUNT_LABEL: 'SAVE_ACCOUNT_LABEL',
  saveAccountLabel: saveAccountLabel,
  AGREE_TO_ETH_WARNING: 'AGREE_TO_ETH_WARNING',
  agreeToEthWarning: agreeToEthWarning,
  SHOW_ETH_WARNING: 'SHOW_ETH_WARNING',
  showEthWarning: showEthWarning,
  // tx conf screen
  COMPLETED_TX: 'COMPLETED_TX',
  TRANSACTION_ERROR: 'TRANSACTION_ERROR',
  NEXT_TX: 'NEXT_TX',
  PREVIOUS_TX: 'PREV_TX',
  setSelectedAddress: setSelectedAddress,
  signMsg: signMsg,
  cancelMsg: cancelMsg,
  sendTx: sendTx,
  signTx: signTx,
  cancelTx: cancelTx,
  completedTx: completedTx,
  txError: txError,
  nextTx: nextTx,
  previousTx: previousTx,
  viewPendingTx: viewPendingTx,
  VIEW_PENDING_TX: 'VIEW_PENDING_TX',
  // app messages
  showAccountDetail: showAccountDetail,
  BACK_TO_ACCOUNT_DETAIL: 'BACK_TO_ACCOUNT_DETAIL',
  backToAccountDetail: backToAccountDetail,
  showAccountsPage: showAccountsPage,
  showConfTxPage: showConfTxPage,
  confirmSeedWords: confirmSeedWords,
  // config screen
  SHOW_CONFIG_PAGE: 'SHOW_CONFIG_PAGE',
  SET_RPC_TARGET: 'SET_RPC_TARGET',
  SET_PROVIDER_TYPE: 'SET_PROVIDER_TYPE',
  USE_ETHERSCAN_PROVIDER: 'USE_ETHERSCAN_PROVIDER',
  useEtherscanProvider: useEtherscanProvider,
  showConfigPage: showConfigPage,
  setRpcTarget: setRpcTarget,
  setProviderType: setProviderType,
  // hacky - need a way to get a reference to account manager
  _setAccountManager: _setAccountManager,
  // loading overlay
  SHOW_LOADING: 'SHOW_LOADING_INDICATION',
  HIDE_LOADING: 'HIDE_LOADING_INDICATION',
  showLoadingIndication: showLoadingIndication,
  hideLoadingIndication: hideLoadingIndication,
  // buy Eth with coinbase
  BUY_ETH: 'BUY_ETH',
  buyEth: buyEth,
}

module.exports = actions

var _accountManager = null
function _setAccountManager (accountManager) {
  _accountManager = accountManager
}

function goHome () {
  return {
    type: actions.GO_HOME,
  }
}

// async actions

function tryUnlockMetamask (password) {
  return (dispatch) => {
    dispatch(actions.unlockInProgress())
    _accountManager.submitPassword(password, (err, selectedAccount) => {
      if (err) {
        dispatch(actions.unlockFailed())
      } else {
        dispatch(actions.unlockMetamask(selectedAccount))
      }
    })
  }
}

function createNewVault (password, entropy) {
  return (dispatch) => {
    dispatch(actions.createNewVaultInProgress())
    _accountManager.createNewVault(password, entropy, (err, result) => {
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }
      dispatch(actions.showNewVaultSeed(result))
    })
  }
}

function revealSeedConfirmation () {
  return {
    type: this.REVEAL_SEED_CONFIRMATION,
  }
}

function requestRevealSeed (password) {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())
    _accountManager.tryPassword(password, (err, seed) => {
      dispatch(actions.hideLoadingIndication())
      if (err) return dispatch(actions.displayWarning(err.message))
      _accountManager.recoverSeed((err, seed) => {
        if (err) return dispatch(actions.displayWarning(err.message))
        dispatch(actions.showNewVaultSeed(seed))
      })
    })
  }
}

function recoverFromSeed (password, seed) {
  return (dispatch) => {
    // dispatch(actions.createNewVaultInProgress())
    dispatch(actions.showLoadingIndication())
    _accountManager.recoverFromSeed(password, seed, (err, metamaskState) => {
      dispatch(actions.hideLoadingIndication())
      if (err) return dispatch(actions.displayWarning(err.message))

      var account = Object.keys(metamaskState.identities)[0]
      dispatch(actions.unlockMetamask(account))
    })
  }
}

function showInfoPage () {
  return {
    type: actions.SHOW_INFO_PAGE,
  }
}

function setSelectedAddress (address) {
  return (dispatch) => {
    _accountManager.setSelectedAddress(address)
  }
}

function revealAccount () {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())
    _accountManager.revealAccount((err) => {
      dispatch(actions.hideLoadingIndication())
      if (err) return dispatch(actions.displayWarning(err.message))
      dispatch({
        type: actions.REVEAL_ACCOUNT,
      })
    })
  }
}

function signMsg (msgData) {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())

    _accountManager.signMessage(msgData, (err) => {
      dispatch(actions.hideLoadingIndication())

      if (err) return dispatch(actions.displayWarning(err.message))
      dispatch(actions.completedTx(msgData.metamaskId))
    })
  }
}

function signTx (txData) {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())

    web3.eth.sendTransaction(txData, (err, data) => {
      dispatch(actions.hideLoadingIndication())

      if (err) return dispatch(actions.displayWarning(err.message))
      dispatch(actions.hideWarning())
      dispatch(actions.goHome())
    })
  }
}

function sendTx (txData) {
  return (dispatch) => {
    _accountManager.approveTransaction(txData.id, (err) => {
      if (err) {
        alert(err.message)
        dispatch(actions.txError(err))
        return console.error(err.message)
      }
      dispatch(actions.completedTx(txData.id))
    })
  }
}

function completedTx (id) {
  return {
    type: actions.COMPLETED_TX,
    id,
  }
}

function txError (err) {
  return {
    type: actions.TRANSACTION_ERROR,
    message: err.message,
  }
}

function cancelMsg (msgData) {
  _accountManager.cancelMessage(msgData.id)
  return actions.completedTx(msgData.id)
}

function cancelTx (txData) {
  _accountManager.cancelTransaction(txData.id)
  return actions.completedTx(txData.id)
}

//
// initialize screen
//

function showCreateVault () {
  return {
    type: actions.SHOW_CREATE_VAULT,
  }
}

function showRestoreVault () {
  return {
    type: actions.SHOW_RESTORE_VAULT,
  }
}

function showInitializeMenu () {
  return {
    type: actions.SHOW_INIT_MENU,
  }
}

function agreeToDisclaimer () {
  return (dispatch) => {
    dispatch(this.showLoadingIndication())
    _accountManager.agreeToDisclaimer((err) => {
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }

      dispatch(this.hideLoadingIndication())
      dispatch({
        type: this.AGREE_TO_DISCLAIMER,
      })
    })
  }
}

function createNewVaultInProgress () {
  return {
    type: actions.CREATE_NEW_VAULT_IN_PROGRESS,
  }
}

function showNewVaultSeed (seed) {
  return {
    type: actions.SHOW_NEW_VAULT_SEED,
    value: seed,
  }
}

//
// unlock screen
//

function unlockInProgress () {
  return {
    type: actions.UNLOCK_IN_PROGRESS,
  }
}

function unlockFailed () {
  return {
    type: actions.UNLOCK_FAILED,
  }
}

function unlockMetamask (account) {
  return {
    type: actions.UNLOCK_METAMASK,
    value: account,
  }
}

function updateMetamaskState (newState) {
  return {
    type: actions.UPDATE_METAMASK_STATE,
    value: newState,
  }
}

function lockMetamask () {
  return (dispatch) => {
    _accountManager.setLocked((err) => {
      dispatch(actions.hideLoadingIndication())
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }

      dispatch({
        type: actions.LOCK_METAMASK,
      })
    })
  }
}

function showAccountDetail (address) {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())
    _accountManager.setSelectedAddress(address, (err, address) => {
      dispatch(actions.hideLoadingIndication())
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }

      dispatch({
        type: actions.SHOW_ACCOUNT_DETAIL,
        value: address,
      })
    })
  }
}

function backToAccountDetail (address) {
  return {
    type: actions.BACK_TO_ACCOUNT_DETAIL,
    value: address,
  }
}
function clearSeedWordCache (account) {
  return {
    type: actions.CLEAR_SEED_WORD_CACHE,
    value: account,
  }
}

function confirmSeedWords () {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())
    _accountManager.clearSeedWordCache((err, account) => {
      dispatch(actions.hideLoadingIndication())
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }

      console.log('Seed word cache cleared. ' + account)
      dispatch(actions.showAccountDetail(account))
    })
  }
}

function showAccountsPage () {
  return {
    type: actions.SHOW_ACCOUNTS_PAGE,
  }
}

function showConfTxPage () {
  return {
    type: actions.SHOW_CONF_TX_PAGE,
  }
}

function nextTx () {
  return {
    type: actions.NEXT_TX,
  }
}

function viewPendingTx (txId) {
  return {
    type: actions.VIEW_PENDING_TX,
    value: txId,
  }
}

function previousTx () {
  return {
    type: actions.PREVIOUS_TX,
  }
}

function showConfigPage (transitionForward = true) {
  return {
    type: actions.SHOW_CONFIG_PAGE,
    value: transitionForward,
  }
}

//
// config
//

function setRpcTarget (newRpc) {
  _accountManager.setRpcTarget(newRpc)
  return {
    type: actions.SET_RPC_TARGET,
    value: newRpc,
  }
}

function setProviderType (type) {
  _accountManager.setProviderType(type)
  return {
    type: actions.SET_PROVIDER_TYPE,
    value: type,
  }
}

function useEtherscanProvider () {
  _accountManager.useEtherscanProvider()
  return {
    type: actions.USE_ETHERSCAN_PROVIDER,
  }
}

function showLoadingIndication () {
  return {
    type: actions.SHOW_LOADING,
  }
}

function hideLoadingIndication () {
  return {
    type: actions.HIDE_LOADING,
  }
}

function showWarning (text) {
  return this.displayWarning(text)
}

function displayWarning (text) {
  return {
    type: actions.DISPLAY_WARNING,
    value: text,
  }
}

function hideWarning () {
  return {
    type: actions.HIDE_WARNING,
  }
}

function requestExportAccount () {
  return {
    type: actions.REQUEST_ACCOUNT_EXPORT,
  }
}

function exportAccount (address) {
  var self = this

  return function (dispatch) {
    dispatch(self.showLoadingIndication())

    _accountManager.exportAccount(address, function (err, result) {
      dispatch(self.hideLoadingIndication())

      if (err) {
        console.error(err)
        return dispatch(self.displayWarning('Had a problem exporting the account.'))
      }

      dispatch(self.showPrivateKey(result))
    })
  }
}

function showPrivateKey (key) {
  return {
    type: actions.SHOW_PRIVATE_KEY,
    value: key,
  }
}

function saveAccountLabel (account, label) {
  return (dispatch) => {
    dispatch(actions.showLoadingIndication())
    _accountManager.saveAccountLabel(account, label, (err) => {
      dispatch(actions.hideLoadingIndication())
      if (err) {
        return dispatch(actions.showWarning(err.message))
      }
      dispatch({
        type: actions.SAVE_ACCOUNT_LABEL,
        value: { account, label },
      })
    })
  }
}

function showSendPage () {
  return {
    type: actions.SHOW_SEND_PAGE,
  }
}

function agreeToEthWarning () {
  return (dispatch) => {
    _accountManager.agreeToEthWarning((err) => {
      if (err) {
        return dispatch(actions.showEthWarning(err.message))
      }
      dispatch({
        type: actions.AGREE_TO_ETH_WARNING,
      })
    })
  }
}

function showEthWarning () {
  return {
    type: actions.SHOW_ETH_WARNING,
  }
}

function buyEth (address, amount) {
  return (dispatch) => {
    _accountManager.buyEth(address, amount)
    dispatch({
      type: actions.BUY_ETH,
    })
  }
}
