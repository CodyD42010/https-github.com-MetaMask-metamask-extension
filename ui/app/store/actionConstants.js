
export default {
  GO_HOME: 'GO_HOME',
  // modal state
  MODAL_OPEN: 'UI_MODAL_OPEN',
  MODAL_CLOSE: 'UI_MODAL_CLOSE',
  // notification state
  CLOSE_NOTIFICATION_WINDOW: 'CLOSE_NOTIFICATION_WINDOW',
  // sidebar state
  SIDEBAR_OPEN: 'UI_SIDEBAR_OPEN',
  SIDEBAR_CLOSE: 'UI_SIDEBAR_CLOSE',
  // alert state
  ALERT_OPEN: 'UI_ALERT_OPEN',
  ALERT_CLOSE: 'UI_ALERT_CLOSE',
  QR_CODE_DETECTED: 'UI_QR_CODE_DETECTED',
  // network dropdown open
  NETWORK_DROPDOWN_OPEN: 'UI_NETWORK_DROPDOWN_OPEN',
  NETWORK_DROPDOWN_CLOSE: 'UI_NETWORK_DROPDOWN_CLOSE',
  // transition state
  TRANSITION_FORWARD: 'TRANSITION_FORWARD',
  // remote state
  UPDATE_METAMASK_STATE: 'UPDATE_METAMASK_STATE',
  FORGOT_PASSWORD: 'FORGOT_PASSWORD',
  CLOSE_WELCOME_SCREEN: 'CLOSE_WELCOME_SCREEN',
  // unlock screen
  UNLOCK_IN_PROGRESS: 'UNLOCK_IN_PROGRESS',
  UNLOCK_FAILED: 'UNLOCK_FAILED',
  UNLOCK_SUCCEEDED: 'UNLOCK_SUCCEEDED',
  LOCK_METAMASK: 'LOCK_METAMASK',
  // error handling
  DISPLAY_WARNING: 'DISPLAY_WARNING',
  HIDE_WARNING: 'HIDE_WARNING',
  // accounts screen
  SET_SELECTED_TOKEN: 'SET_SELECTED_TOKEN',
  SHOW_ACCOUNT_DETAIL: 'SHOW_ACCOUNT_DETAIL',
  SHOW_ACCOUNTS_PAGE: 'SHOW_ACCOUNTS_PAGE',
  SHOW_CONF_TX_PAGE: 'SHOW_CONF_TX_PAGE',
  SET_CURRENT_FIAT: 'SET_CURRENT_FIAT',
  // account detail screen
  SHOW_SEND_TOKEN_PAGE: 'SHOW_SEND_TOKEN_PAGE',
  SHOW_PRIVATE_KEY: 'SHOW_PRIVATE_KEY',
  SET_ACCOUNT_LABEL: 'SET_ACCOUNT_LABEL',
  SET_NETWORK_NONCE: 'SET_NETWORK_NONCE',
  // tx conf screen
  COMPLETED_TX: 'COMPLETED_TX',
  TRANSACTION_ERROR: 'TRANSACTION_ERROR',
  UPDATE_TRANSACTION_PARAMS: 'UPDATE_TRANSACTION_PARAMS',
  SET_NEXT_NONCE: 'SET_NEXT_NONCE',
  // send screen
  UPDATE_GAS_LIMIT: 'UPDATE_GAS_LIMIT',
  UPDATE_GAS_PRICE: 'UPDATE_GAS_PRICE',
  UPDATE_GAS_TOTAL: 'UPDATE_GAS_TOTAL',
  UPDATE_SEND_HEX_DATA: 'UPDATE_SEND_HEX_DATA',
  UPDATE_SEND_TOKEN_BALANCE: 'UPDATE_SEND_TOKEN_BALANCE',
  UPDATE_SEND_TO: 'UPDATE_SEND_TO',
  UPDATE_SEND_AMOUNT: 'UPDATE_SEND_AMOUNT',
  UPDATE_SEND_ERRORS: 'UPDATE_SEND_ERRORS',
  UPDATE_MAX_MODE: 'UPDATE_MAX_MODE',
  UPDATE_SEND: 'UPDATE_SEND',
  CLEAR_SEND: 'CLEAR_SEND',
  GAS_LOADING_STARTED: 'GAS_LOADING_STARTED',
  GAS_LOADING_FINISHED: 'GAS_LOADING_FINISHED',
  UPDATE_SEND_ENS_RESOLUTION: 'UPDATE_SEND_ENS_RESOLUTION',
  UPDATE_SEND_ENS_RESOLUTION_ERROR: 'UPDATE_SEND_ENS_RESOLUTION_ERROR',
  // config screen
  SET_RPC_TARGET: 'SET_RPC_TARGET',
  SET_PROVIDER_TYPE: 'SET_PROVIDER_TYPE',
  SET_PREVIOUS_PROVIDER: 'SET_PREVIOUS_PROVIDER',
  UPDATE_TOKENS: 'UPDATE_TOKENS',
  SET_HARDWARE_WALLET_DEFAULT_HD_PATH: 'SET_HARDWARE_WALLET_DEFAULT_HD_PATH',
  // loading overlay
  SHOW_LOADING: 'SHOW_LOADING_INDICATION',
  HIDE_LOADING: 'HIDE_LOADING_INDICATION',
  // buy Eth with coinbase
  BUY_ETH: 'BUY_ETH',

  TOGGLE_ACCOUNT_MENU: 'TOGGLE_ACCOUNT_MENU',

  SET_USE_BLOCKIE: 'SET_USE_BLOCKIE',
  SET_USE_NONCEFIELD: 'SET_USE_NONCEFIELD',
  UPDATE_CUSTOM_NONCE: 'UPDATE_CUSTOM_NONCE',
  SET_IPFS_GATEWAY: 'SET_IPFS_GATEWAY',

  SET_PARTICIPATE_IN_METAMETRICS: 'SET_PARTICIPATE_IN_METAMETRICS',
  SET_METAMETRICS_SEND_COUNT: 'SET_METAMETRICS_SEND_COUNT',

  // locale
  SET_CURRENT_LOCALE: 'SET_CURRENT_LOCALE',

  // Feature Flags
  UPDATE_FEATURE_FLAGS: 'UPDATE_FEATURE_FLAGS',

  // Preferences
  UPDATE_PREFERENCES: 'UPDATE_PREFERENCES',

  // Onboarding
  COMPLETE_ONBOARDING: 'COMPLETE_ONBOARDING',

  SET_MOUSE_USER_STATE: 'SET_MOUSE_USER_STATE',

  // Network
  SET_PENDING_TOKENS: 'SET_PENDING_TOKENS',
  CLEAR_PENDING_TOKENS: 'CLEAR_PENDING_TOKENS',

  SET_FIRST_TIME_FLOW_TYPE: 'SET_FIRST_TIME_FLOW_TYPE',

  SET_SELECTED_SETTINGS_RPC_URL: 'SET_SELECTED_SETTINGS_RPC_URL',
  SET_NETWORKS_TAB_ADD_MODE: 'SET_NETWORKS_TAB_ADD_MODE',

  LOADING_METHOD_DATA_STARTED: 'LOADING_METHOD_DATA_STARTED',
  LOADING_METHOD_DATA_FINISHED: 'LOADING_METHOD_DATA_FINISHED',

  LOADING_TOKEN_PARAMS_STARTED: 'LOADING_TOKEN_PARAMS_STARTED',
  LOADING_TOKEN_PARAMS_FINISHED: 'LOADING_TOKEN_PARAMS_FINISHED',

  SET_REQUEST_ACCOUNT_TABS: 'SET_REQUEST_ACCOUNT_TABS',
  SET_CURRENT_WINDOW_TAB: 'SET_CURRENT_WINDOW_TAB',
  SET_OPEN_METAMASK_TAB_IDS: 'SET_OPEN_METAMASK_TAB_IDS',
}
