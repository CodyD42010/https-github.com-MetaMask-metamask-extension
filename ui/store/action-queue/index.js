import pify from 'pify';
import { isManifestV3 } from '../../../shared/modules/mv3.utils';

// // A simplified pify maybe?
// function pify(apiObject) {
//   return Object.keys(apiObject).reduce((promisifiedAPI, key) => {
//     if (apiObject[key].apply) { // depending on our browser support we might use a nicer check for functions here
//       promisifiedAPI[key] = function (...args) {
//         return new Promise((resolve, reject) => {
//           return apiObject[key](
//             ...args,
//             (err, result) => {
//               if (err) {
//                 reject(err);
//               } else {
//                 resolve(result);
//               }
//             },
//           );
//         });
//       };
//     }
//     return promisifiedAPI;
//   }, {});
// }

let background = null;
let promisifiedBackground = null;

const actionRetryQueue = [];

export function __TEST_CLEAR_QUEUE() {
  actionRetryQueue.length = 0;
}

// add action to queue
const addToActionQueue = (actionId, request, resolve, reject) => {
  if (actionRetryQueue.some((act) => act.actionId === actionId)) {
    return;
  }
  actionRetryQueue.push({
    actionId,
    request,
    resolve,
    reject,
  });

  processActionRetryQueue();
};

// remove action from queue on successful completion
// TODO: it should be possible to refactor the way the queue is drained to make this splicing unnecessary but I ran out of time.
const removeFromActionQueue = (actionId) => {
  const index = actionRetryQueue.find((act) => act.actionId === actionId);
  actionRetryQueue.splice(index, 1);
};

// invokes promisifiedBackground method in MV2 content
// In MV3 context the execution is:
//   1. action is added to retry queue, along with resolve handler to be executed on completion of action
//   2. the queue is then immediately processed
//   3. on completion of action either successfully or by throwing exception, action is removed from retry queue
export const submitRequestToBackground = (
  method,
  args = [],
  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
) => {
  if (isManifestV3()) {
    return new Promise((resolve, reject) => {
      addToActionQueue(actionId, { method, args }, resolve, reject);
    });
  }
  return promisifiedBackground[method](...args);
};

// Function is similar to submitRequestToBackground function above
// except that it invokes method on background and not promisifiedBackground
export const callBackgroundMethod = (
  method,
  args = [],
  callback,
  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
) => {
  if (isManifestV3()) {
    const resolve = (value) => callback(null, value);
    const reject = (err) => callback(err);
    addToActionQueue(actionId, { method, args }, resolve, reject);
  } else {
    background[method](...args, callback);
  }
};

// A thenable like Promise.resolve() but it doesn't introduce an asynchronous delay.
const syncThenable = {
  then: (cb) => cb(),
};

// Clears list of pending action in actionRetryQueue
// The results of background calls are wired up to the original promises that's been returned
// The first method on the queue gets called synchronously to make testing and reasoning about
//  a single request to an open connection easier.
function processActionRetryQueue() {
  if (background.connectionStream.readable) {
    // Iterating by index over a queue that's potentially being spliced and pushed to is not great. Let's copy.
    const actionRetryQueueSnapshot = [...actionRetryQueue];
    actionRetryQueueSnapshot.reduce(
      (
        previousPromise,
        { request: { method, args }, actionId, resolve, reject },
      ) => {
        // eslint-disable-next-line consistent-return
        return previousPromise.then(() => {
          if (background.connectionStream.readable) {
            try {
              return promisifiedBackground[method](...args)
                .then((result) => {
                  removeFromActionQueue(actionId);
                  resolve(result);
                })
                .catch((err) => {
                  removeFromActionQueue(actionId);
                  reject(err);
                });
            } catch (err) {
              removeFromActionQueue(actionId);
              reject(err);
            }
          }
          return syncThenable;
        });
      },
      syncThenable,
    );
  }
}

export async function _setBackgroundConnection(backgroundConnection) {
  background = backgroundConnection;
  promisifiedBackground = pify(background);
  if (isManifestV3()) {
    // This function call here will clear the queue of actions
    // collected while connection stream is not available.
    processActionRetryQueue();
  }
}
