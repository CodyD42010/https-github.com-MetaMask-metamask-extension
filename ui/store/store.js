import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'remote-redux-devtools';
import rootReducer from '../ducks';

const logger = createLogger({
  collapsed: true,
});

export default function configureStore(initialState) {
  const composeEnhancers = composeWithDevTools({
    name: 'MetaMask',
    hostname: 'localhost',
    port: 8000,
    realtime: Boolean(process.env.METAMASK_DEBUG),
  });
  return createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(thunkMiddleware, logger)),
  );
}
