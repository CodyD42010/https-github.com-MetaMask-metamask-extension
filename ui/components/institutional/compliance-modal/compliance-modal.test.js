import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import sinon from 'sinon';
import { hideModal } from '../../../store/actions';
import ComplianceModal from '.';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('../../../store/actions', () => ({
  hideModal: jest.fn(),
}));

describe('ComplianceModal', () => {
  let dispatchMock;

  beforeEach(() => {
    dispatchMock = jest.fn();
    useDispatch.mockReturnValue(dispatchMock);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render the correct content', () => {
    const { container, getByTestId } = render(<ComplianceModal />);

    expect(getByTestId('compliance-info')).toBeInTheDocument();
    expect(getByTestId('compliance-bullets')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should close the modal when close button is clicked', () => {
    const { getByTestId } = render(<ComplianceModal />);

    fireEvent.click(getByTestId('compliance-modal-close'));

    expect(hideModal).toHaveBeenCalled();
    expect(dispatchMock).toHaveBeenCalledWith(hideModal());
  });

  it('should open the Compliance page when submit button is clicked', () => {
    global.platform = { openTab: sinon.spy() };
    const { container } = render(<ComplianceModal />);

    const btn = container.getElementsByClassName('btn-primary')[0];

    fireEvent.click(btn);

    expect(global.platform.openTab.called).toBeTruthy();
  });
});
