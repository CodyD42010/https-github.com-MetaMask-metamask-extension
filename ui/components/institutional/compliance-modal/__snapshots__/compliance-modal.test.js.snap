// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComplianceModal should render the correct content 1`] = `
<div>
  <div
    class="modal-container"
  >
    <div
      class="modal-container__content"
    >
      <div
        class="box box--flex-direction-row"
      >
        <header
          class="box box--display-flex box--flex-direction-row box--justify-content-space-between"
        >
          <div
            class="box box--display-flex box--flex-direction-row box--justify-content-flex-start box--align-items-center"
          >
            <img
              alt="Codefi Compliance"
              height="32"
              src="images/compliance-logo-small.svg"
              width="32"
            />
            <h4
              class="box mm-text mm-text--body-md box--margin-left-2 box--flex-direction-row box--color-text-default"
            >
              [codefiCompliance]
            </h4>
          </div>
          <button
            aria-label="[close]"
            class="box mm-button-icon mm-button-icon--size-sm box--display-inline-flex box--flex-direction-row box--justify-content-center box--align-items-center box--color-icon-default box--background-color-transparent box--rounded-lg"
            data-testid="compliance-modal-close"
          >
            <span
              class="box mm-icon mm-icon--size-sm box--display-inline-block box--flex-direction-row box--color-inherit"
              style="mask-image: url('./images/icons/close.svg');"
            />
          </button>
        </header>
        <p
          class="box mm-text mm-text--body-md box--padding-bottom-3 box--flex-direction-row box--color-text-alternative"
          data-testid="compliance-info"
        >
          [complianceBlurb0]
        </p>
        <p
          class="box mm-text mm-text--body-md box--padding-bottom-3 box--flex-direction-row box--color-text-alternative"
        >
          [complianceBlurb1]
        </p>
        <p
          class="box mm-text mm-text--body-md box--padding-bottom-3 box--flex-direction-row box--color-text-alternative"
        >
          [complianceBlurpStep0]
        </p>
        <ol
          data-testid="compliance-bullets"
        >
          <li>
            [complianceBlurbStep1]
          </li>
          <li>
            [complianceBlurbStep2]
          </li>
          <li>
            [complianceBlurbStep3]
          </li>
          <li>
            [complianceBlurbStep4]
          </li>
          <li>
            [complianceBlurbStep5]
          </li>
        </ol>
      </div>
    </div>
    <div
      class="modal-container__footer"
    >
      <button
        class="button btn--rounded btn-primary modal-container__footer-button"
        role="button"
        tabindex="0"
      >
        [openCodefiCompliance]
      </button>
    </div>
  </div>
</div>
`;
